web/src/engine/keyboard/src/keyboards/keyboardLoaderBase.ts

```diff
@@ -76,23 +16,49 @@ export default abstract class KeyboardLoaderBase {
     this._harness = harness;
   }
 
-  public loadKeyboardFromPath(uri: string): Promise<Keyboard> {
+  /**
+   * Load a keyboard from a remote or local URI.
+   *
+   * @param uri  The URI of the keyboard to load.
+   * @returns    A Promise that resolves to the loaded keyboard.
+   */
+  public async loadKeyboardFromPath(uri: string): Promise<Keyboard> {
     this.harness.install();
-    const promise = this.loadKeyboardInternal(uri, new UriBasedErrorBuilder(uri));
-
-    return promise;
+    return this.loadKeyboardInternal(uri, new UriBasedErrorBuilder(uri));
   }
 
-  public loadKeyboardFromStub(stub: KeyboardStub) {
+  /**
+   * Load a keyboard from keyboard stub.
+   *
+   * @param stub  The stub of the keyboard to load.
+   * @returns     A Promise that resolves to the loaded keyboard.
+   */
+  public async loadKeyboardFromStub(stub: KeyboardStub): Promise<Keyboard> {
     this.harness.install();
-    let promise = this.loadKeyboardInternal(stub.filename, new StubBasedErrorBuilder(stub), stub.id);
+    return this.loadKeyboardInternal(stub.filename, new StubBasedErrorBuilder(stub));
+  }
+
+  private async loadKeyboardInternal(uri: string, errorBuilder: KeyboardLoadErrorBuilder): Promise<Keyboard> {
+    const blob = await this.loadKeyboardBlob(uri, errorBuilder);
 
-    return promise;
+    let script: string;
+    try {
+      script = await blob.text();
+    } catch (e) {
+      throw errorBuilder.invalidKeyboard(e);
+    }
+
+    if (script.startsWith('KXTS', 0)) {
+      // KMX or LDML (KMX+) keyboard
+      console.error("KMX keyboard loading is not yet implemented!");
+      return null;
+    }
```

Comment:
I just realised that this is not going to work with .kmx, because that's a binary format and so it will fail the `.text()` conversion which expects UTF-8. We need to test `blob.slice(0,4).bytes() == 'KXTS'` (with the appropriate Uint8Array comparison which I leave to you) and only attempt to convert to UTF-8 if it's not a .kmx.

Reply:
Note that `Blob` requires relatively recent browsers. I haven't checked if this pushes our minimum versions -- can you verify?

https://developer.mozilla.org/en-US/docs/Web/API/Blob/text#browser_compatibility

Reply:
MacOS doesn't matter because we aren't tied to Safari version on macOS. It's only relevant for iOS.